ceremonies/meetup-validation/src/lib.rs

```diff
@@ -0,0 +1,253 @@
+					filter_attestations(&participants_to_process, &relevant_attestations);
+				continue
+			}
+		} else {
+			if min_num_outgoing_attestations < threshold_fn(participants_to_process.len()) {
+				let ps = participants_grouped_by_outgoing_attestations[0].1.clone();
+				ps.clone().into_iter().for_each(|p| {
+					excluded_participants.push((p, ExclusionReason::TooFewOutgoingAttestations))
+				});
+
+				// remove the participants from the included participants and the attestation vectors
+				participants_to_process.retain(|k| !ps.contains(k));
+				relevant_attestations =
+					filter_attestations(&participants_to_process, &relevant_attestations);
+				continue
+			}
+		}
+
+		// if all participants are above the threshold and therefore no participants were removed, we exit the loop
+		break
+	}
+	excluded_participants
+}
+
+fn find_majority_vote(
+	participants: &Vec<usize>,
+	participant_votes: &Vec<u32>,
+) -> Result<(u32, u32), MeetupValidationError> {
+	let mut n_vote_candidates: Vec<(u32, u32)> = vec![];
+	for i in participants {
+		let this_vote = participant_votes[*i];
+		match n_vote_candidates.iter().position(|&(n, _c)| n == this_vote) {
+			Some(idx) => n_vote_candidates[idx].1 += 1, //safe; <= number of candidates
+			_ => n_vote_candidates.insert(0, (this_vote, 1)),
+		};
+	}
+
+	if n_vote_candidates.is_empty() {
+		return Err(MeetupValidationError::BallotEmpty)
+	}
+	// sort by descending vote count
+	n_vote_candidates.sort_by(|a, b| b.1.cmp(&a.1));
+	if n_vote_candidates[0].1 < 3 {
+		//safe; n_vote_candidate not empty checked aboveÏ
+		return Err(MeetupValidationError::NoDependableVote)
+	}
+	let (n_confirmed, vote_count) = n_vote_candidates[0];
+	Ok((n_confirmed, vote_count))
+}
+
+fn filter_attestations(
+	participants: &Vec<usize>,
+	participant_attestations: &Vec<Vec<usize>>,
+) -> Vec<Vec<usize>> {
+	// filter out participants from the attestation vectors that are not in the participants vector anymore.
+	participant_attestations
+		.clone()
+		.iter()
+		.map(|a| a.clone().into_iter().filter(|j| participants.contains(j)).collect())
+		.collect()
+}
+
+fn group_participants_by_num_incoming_and_outgoing_attestations(
+	participants: &Vec<usize>,
+	participant_attestations: &Vec<Vec<usize>>,
+) -> (Vec<(usize, Vec<usize>)>, Vec<(usize, Vec<usize>)>) {
+	let num_outgoing_attestations: Vec<usize> =
+		participant_attestations.iter().map(|a| a.len()).collect();
+	let num_incoming_attestations: Vec<usize> = (0..participant_attestations.len())
+		.into_iter()
+		.map(|p| {
+			participant_attestations
+				.into_iter()
+				.enumerate()
+				.filter(|(idx, a)| &p != idx && a.contains(&p))
+				.map(|item| item.1.clone())
+				.collect::<Vec<Vec<usize>>>()
+				.len()
+		})
+		.collect();
+	return (
+		group_participants_by_num_attestations(participants, &num_outgoing_attestations),
+		group_participants_by_num_attestations(participants, &num_incoming_attestations),
+	)
+}
+
+fn group_participants_by_num_attestations(
```

Comment:
your naming is more verbose than necessary: for this fn it doesn't matter that we're treating `participants` and `attestations`. It just takes in index and a number per index and sorts and groups. I'd prefer to just describe what the function does, not what we are using it for